### PR TITLE
x.py: run `rustup toolchain link` in setup

### PR DESCRIPTION
**Pull request by Sl1mb0**
Addresses #89206  

r? @jyn514

### CODE REVIEW

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:115]: This is checking the wrong thing - `output()` only fails if the syscall itself fails, not if the status is non-zero.
```suggestion
    if toolchain_link.is_err() || !toolchain_link.status().success() {
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:112]: You need to create this directory first, otherwise it will give an error:
```
$ rustup toolchain link stage1 build/x86_64-unknown-linux-gnu/stage1
error: not a directory: 'build/x86_64-unknown-linux-gnu/stage1/lib'
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:114]: I just realized this will overwrite any existing toolchain - that seems kind of not great, it means we'll delete any existing toolchain with this name. @kinnison @rbtcollins is that intentional that it will silently overwrite?

I guess to avoid that we could explicitly run `rustup toolchain list` and see if it already has `stage1`.

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:122]: It would be nice to let people know x.py did this, so they know to use it in the future.
```suggestion
    } else {
        println!("Added `stage1` rustup toolchain; use `cargo +stage1 build` to run a newly-built toolchain");
    }
```

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/bootstrap/setup.rs
Comment (workingjubilee) [inline src/bootstrap/setup.rs:120]: So... I could have missed something about the previous code, but at this point in the program, what proof do we have that the user **has** Rustup on their system that justifies emitting this message all "WARNING!" style, telling them about a binary that possibly isn't installed on their system? It is entirely possible that the user is bootstrapping Rust on their system for the first time, Rustup is not available for this target, and we should be careful to neither alarm nor distract them. Something gentler and softer-voiced seems appropriate.

File: src/bootstrap/setup.rs
Comment (Sl1mb0) [reply src/bootstrap/setup.rs:112]: Do you think we should also check if `rustup` has been installed and inform the user if it hasn't?

File: src/bootstrap/setup.rs
Comment (workingjubilee) [reply src/bootstrap/setup.rs:122]: Perhaps a command that would be less disruptive if it was executed immediately on the repository they are in, which could cause amusing errors? Or perhaps additional context to when and where they should run it. There are actually some people who first try Rust by building from the source, so we shouldn't assume they are an old hand.

File: src/bootstrap/setup.rs
Comment (workingjubilee) [reply src/bootstrap/setup.rs:114]: Usually this command is executed by the user, so like most Unix-y commands, dispatches as silently and swiftly as possible, and alerting someone is a sign of failure.

**Review by workingjubilee — COMMENTED** [review]
Have we considered the possible failure mode of a distro-installed Rustup that is later overshadowed by Rustup's normal `/home/user/.{cargo,rustup}` install process? Possibly that doesn't matter: maybe such a case would make one expect possible overshadowing.

File: src/bootstrap/setup.rs
Comment (jyn514) [reply src/bootstrap/setup.rs:114]: Hmm - but `ln -s` (which you said in the original issue is basically what `toolchain link` does) gives an error if the file already exists:
```
$ ln -s x config.toml
ln: failed to create symbolic link 'config.toml': File exists
```

File: src/bootstrap/setup.rs
Comment (Sl1mb0) [reply src/bootstrap/setup.rs:120]: If rustup is installed:
```
Linking stage1 toolchain
```

If not:
```
rustup is not installed
Unable to link stage1 toolchain
```

Comment (jyn514) [reply src/bootstrap/setup.rs:120]: I think it would be fine to only warn if the profile isn't `"user"` (which is what distros should be using); for everyone else they would find rustup really useful for rust-lang/rust even if they wouldn't use it normally.

File: src/bootstrap/setup.rs
Comment (jyn514) [reply src/bootstrap/setup.rs:122]: > Perhaps a command that would be less disruptive if it was executed immediately on the repository they are in, which could cause amusing errors?

Ideally we would fix this with https://github.com/rust-lang/rust/issues/76444 ... but I agree in the meantime we should suggest something else. Maybe add "on a separate rust project" after the `cargo build` snippet?

**Comment by jyn514** [comment]
> Have we considered the possible failure mode of a distro-installed Rustup that is later overshadowed by Rustup's normal /home/user/.{cargo,rustup} install process?

I'm not sure how this could go wrong. If there's a distro-installed rustup, it could run `toolchain link` just like a binary-installed rustup. `toolchain link` doesn't change the default, just add additional toolchains.

File: src/bootstrap/setup.rs
Comment (workingjubilee) [reply src/bootstrap/setup.rs:114]: Reasonable~
But the other reason I mentioned links is due to problems due to OS control over when files may be edited/clobbered/removed, which has less to do with the particular link.

File: src/bootstrap/setup.rs
Comment (Sl1mb0) [reply src/bootstrap/setup.rs:122]: Could change `use` to `try`

File: src/bootstrap/setup.rs
Comment (jyn514) [reply src/bootstrap/setup.rs:112]: Sure, doing that if the profile isn't "user" seems helpful :)

**Comment by workingjubilee** [comment]
I more meant user surprisal due to the overshadowing and the linked toolchain "disappearing", if that mattered.

**Comment by jyn514** [comment]
I clarified offline - the way this can go wrong is:
1. Someone has a version of rustup installed from their distro (in /usr/bin/rustup or wherever).
2. They build rust from source for *reasons* and x.py setup runs `rustup toolchain link stage1`.
3. They install a new version of rustup so they can download new toolchains.
4. The linked toolchain disappears.

Given that in this (somewhat convoluted scenario) they'll be able to easily rerun `toolchain link` or just run `rustup install`, I'm ok with the toolchain disappearing in that case; I don't see a way to avoid it and the benefit here massively outweighs the risk IMO.

File: src/bootstrap/setup.rs
Comment (workingjubilee) [reply src/bootstrap/setup.rs:120]: It's mostly about framing: gently observing that they do so is fine, as long as it doesn't appear on every invocation after, if they decide they don't want to use it.

Comment (jyn514) [reply src/bootstrap/setup.rs:120]: @workingjubilee yup - this is only running once when you first run `setup`, it's never repeated.

File: src/bootstrap/setup.rs
Comment (Sl1mb0) [reply src/bootstrap/setup.rs:114]: Since we have to create the directory in the first place, couldn't we just not link if it already exists?

Comment (jyn514) [reply src/bootstrap/setup.rs:114]: > Since we have to create the directory in the first place, couldn't we just not link if it already exists?

That doesn't seem great - that won't warn or setup the toolchain if the user has already run `x.py build` but not created a rustup toolchain.

Comment (Sl1mb0) [reply src/bootstrap/setup.rs:114]: I think this is why @Mark-Simulacrum suggested using a stamp file. We could also extend this by adding configuration for it in the `config.toml`; e.g. `toolchain-overwrite` or something where the user can specify if they want to overwrite the toolchain every time they run `x.py setup` or not, and by default `./x.py setup` will only do it once.

Comment (jyn514) [reply src/bootstrap/setup.rs:114]: I really don't think this needs additional configuration. We can just check if `rustup toolchain list` contains `stage1`.

File: src/bootstrap/setup.rs
Comment (Sl1mb0) [reply src/bootstrap/setup.rs:112]: Is there a best practice for getting this information?

Comment (jyn514) [reply src/bootstrap/setup.rs:112]: The profile is passed in as an argument to `setup`.

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:118]: This is not right. It needs to check if rustup is anywhere in PATH; and ideally if it's executable. Try seeing if `rustup --version` succeeds or something like that.

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:126]:
```suggestion
        if let Ok(toolchain_list) = toolchain_list {
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:152]: I don't think saying "rustup" is correct here - it's x.py that's making the decision. But the output comes from x.py already so I don't think we need to mention that explicitly.
```suggestion
                "`rustup` failed to list the current toolchains; not attempting to link `stage1` toolchain"
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:147]: ditto here, saying "rustup will not attempt" is not right

**Review by jyn514 — APPROVED** [review]

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:118]: This needs to check `status()`, remember.

Comment (Sl1mb0) [reply src/bootstrap/setup.rs:118]: 🤦

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:117]:
```suggestion
    let rustup_installed = rustup_installed.map_or(false, |status| status.success());
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:163]: This long list of `else` blocks makes it hard to tell which error corresponds with which check IMO - could you instead use a separate functions and early returns so it's more clear? Something like this:
```rust
    maybe_create_rustup_toolchain();
// ...
}

fn maybe_create_rustup_toolchain() {
     if !rustup_installed || !stage_dir_exists {
        // ...
        return
     }
     // ...
}
```

File: src/bootstrap/setup.rs
Comment (jyn514) [inline src/bootstrap/setup.rs:124]: IMO the current output is a little hard to read - maybe add a newline before any possible output?
```
Please choose one (a/b/c/d/e): a
`x.py` will now use the configuration at /home/joshua/src/rust/rust/src/bootstrap/defaults/config.library.toml
Added `stage1` rustup toolchain; try `cargo +stage1 build` on a separate rust project to run a newly-built toolchain

Rust's CI will automatically fail if it doesn't pass `tidy`, the internal tool for ensuring code quality.
```
```suggestion
println!();
```

**Review by jyn514 — COMMENTED** [review]
Looks pretty good to me :) I just have a couple last nits.

@Mark-Simulacrum do you want to take a look before I approve? No problem if not, this seems pretty low risk.

**Comment by Mark-Simulacrum** [comment]
No, happy to let you own the review here. Can always adjust later in future PRs if it becomes necessary :)

**Comment by jyn514** [comment]
@bors r+

This is awesome, thank you! :heart:

**Comment by Mark-Simulacrum** [comment]
@bors r-

Please squash the commits here to keep history a little cleaner.

**Pull request merged**
